crypto/crypto/src/multi_ed25519.rs

```diff
@@ -0,0 +1,547 @@
+            .try_for_each(|sig_bytes| deserialize_and_push(sig_bytes, &mut signatures));
+
+        multi_signature.map(|_| MultiEd25519Signature(signatures, bitvec))
+    }
+}
+
+impl Length for MultiEd25519Signature {
+    fn length(&self) -> usize {
+        self.to_bytes().len()
+    }
+}
+
+impl std::hash::Hash for MultiEd25519Signature {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        let encoded_signature = self.to_bytes();
+        state.write(&encoded_signature);
+    }
+}
+
+// Those are required by the implementation of hash above.
+impl PartialEq for MultiEd25519Signature {
+    fn eq(&self, other: &MultiEd25519Signature) -> bool {
+        self.to_bytes() == other.to_bytes()
+    }
+}
+
+impl Eq for MultiEd25519Signature {}
+
+impl fmt::Display for MultiEd25519Signature {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", hex::encode(&self.to_bytes()[..]))
+    }
+}
+
+impl fmt::Debug for MultiEd25519Signature {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "MultiEd25519Signature({})", self)
+    }
+}
+
+impl ValidKey for MultiEd25519Signature {
+    fn to_bytes(&self) -> Vec<u8> {
+        self.to_bytes()
+    }
+}
+
+//////////////////////
+// Signature Traits //
+//////////////////////
+
+impl Signature for MultiEd25519Signature {
+    type VerifyingKeyMaterial = MultiEd25519PublicKey;
+    type SigningKeyMaterial = MultiEd25519PrivateKey;
+
+    /// Checks that `self` is valid for `message` using `public_key`.
+    fn verify(&self, message: &HashValue, public_key: &MultiEd25519PublicKey) -> Result<()> {
+        self.verify_arbitrary_msg(message.as_ref(), public_key)
+    }
+
+    /// Checks that `self` is valid for an arbitrary &[u8] `message` using `public_key`.
+    /// Outside of this crate, this particular function should only be used for native signature
+    /// verification in Move.
+    fn verify_arbitrary_msg(
+        &self,
+        message: &[u8],
+        public_key: &MultiEd25519PublicKey,
+    ) -> Result<()> {
+        if last_bit_set(&self.1) > public_key.length()
+            || count_set_bits(&self.1) < public_key.1 as usize
+        {
+            return Err(anyhow!("{}", CryptoMaterialError::ValidationError));
+        }
+        let mut bitvec_index = 0;
+        // TODO use deterministic batch verification
+        for sig in &self.0 {
+            while !self.1[bitvec_index] {
+                bitvec_index += 1;
+            }
+            sig.verify_arbitrary_msg(message, &public_key.0[bitvec_index])?;
+            bitvec_index += 1;
+        }
+        Ok(())
+    }
+
+    fn to_bytes(&self) -> Vec<u8> {
+        self.to_bytes()
+    }
+}
+
+//////////////////////
+// Helper functions //
+//////////////////////
+
+// Helper function required to MultiEd25519 keys to_bytes to add or omit the threshold.
+fn to_bytes<T: ValidKey>(keys: &[T], threshold: u8) -> Vec<u8> {
```

Comment:
Same comment as above: 
1. you can `flat_map` the vecs
2. unsure about the best way to save space here.